File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
# 맵의 상태 : maps, 0 : 벽, 1 : 벽 x 
# 캐릭터가 상대 팀 진영에 도착하기 위해서 지나가야 하는 칸의 개수의 최솟값
# 상대 팀 진영에 도착할 수 없을 때 : -1
# bfs
from collections import deque
def solution(maps):
    n = len(maps)
    m = len(maps[0])
    
    directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
    
    q = deque([(0, 0, 1)]) # row, col, path_length
    visited = set()
    visited.add((0, 0))
    
    while q:
        row, col, path_length = q.popleft()
        
        if row == n-1 and col == m-1:
            return path_length
        
        for dr, dc in directions:
            new_row = row + dr
            new_col = col + dc
            
            if 0 <= new_row < n and 0<= new_col < m and maps[new_row][new_col] == 1:
                if (new_row, new_col) not in visited:
                    q.append((new_row, new_col, path_length + 1))
                    visited.add((new_row, new_col))
        
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    // 최솟값 -> bfs(queue)
    public static ArrayDeque<Position> q = new ArrayDeque<>();
    public static int[][] directions = new int[][]{{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    
    public int solution(int[][] maps) {
        int n = maps.length, m = maps[0].length;
        int row = 0, col = 0;
        
        q.addLast(new Position(row, col, 1));
        return bfs(maps, n, m);
    }
    
    public int bfs(int[][] maps, int n, int m) {
        while(!q.isEmpty()) {
            Position p = q.pollFirst();
            
            if (p.row == n-1 && p.col == m-1) return p.path;
            
            for (int[] dir : directions) {
                int new_row = p.row + dir[0];
                int new_col = p.col + dir[1];
                
                if (0 <= new_row && new_row < n 
                    && 0 <= new_col && new_col < m
                    && maps[new_row][new_col] == 1) {
                    maps[new_row][new_col] = 0;
                    q.addLast(new Position(new_row, new_col, p.path + 1));
                }
            }
        }
        
        return -1;
    }
    
    class Position {
        int row, col, path;
        
        Position(int row, int col, int path) {
            this.row = row;
            this.col = col;
            this.path = path;
        }
    }
}